File: java/src/main/java/com/acme/wealth/EntityContext.java
package com.acme.wealth;

import org.yaml.snakeyaml.Yaml;

import java.io.InputStream;
import java.math.BigDecimal;
import java.util.List;
import java.util.Map;

/**
 * Immutable entity context loaded once at process start from ACME_ENTITY_ID env var.
 * Defaults to SG when the variable is absent (baseline preservation).
 */
public final class EntityContext {

    private static final EntityContext INSTANCE = load();

    private final String entityId;
    private final String currency;
    private final String locale;
    private final String regulator;
    private final String bookingCentre;
    private final BigDecimal feeBps;
    private final BigDecimal largePositionThreshold;
    private final String suitabilityFramework;
    private final String kmsVaultName;
    private final String dataRegion;
    private final BigDecimal minBalance;
    private final String disclosureLocale;
    private final String auditLogPolicy;
    private final String dbConnectionSecret;
    private final String apiKeySecret;

    private EntityContext(Map<String, Object> cfg) {
        this.entityId = str(cfg, "entity_id");
        this.currency = str(cfg, "currency");
        this.locale = str(cfg, "locale");
        this.regulator = str(cfg, "regulator");
        this.bookingCentre = str(cfg, "booking_centre");
        this.feeBps = decimal(cfg, "fee_bps");
        this.largePositionThreshold = decimal(cfg, "large_position_threshold");
        this.suitabilityFramework = str(cfg, "suitability_framework");
        this.kmsVaultName = str(cfg, "kms_vault_name");
        this.dataRegion = str(cfg, "data_region");
        this.minBalance = decimal(cfg, "min_balance");
        this.disclosureLocale = str(cfg, "disclosure_locale");
        this.auditLogPolicy = str(cfg, "audit_log_policy");
        // Secrets via Azure Key Vault references — never plaintext
        String vault = this.kmsVaultName;
        this.dbConnectionSecret = "${KEYVAULT:" + vault + "-db-connection-string}";
        this.apiKeySecret = "${KEYVAULT:" + vault + "-api-key}";
    }

    @SuppressWarnings("unchecked")
    private static EntityContext load() {
        String entityId = System.getenv("ACME_ENTITY_ID");
        if (entityId == null || entityId.isBlank()) {
            entityId = "SG";
        }
        Yaml yaml = new Yaml();
        try (InputStream is = EntityContext.class.getClassLoader()
                .getResourceAsStream("entity_config.yaml")) {
            if (is == null) {
                throw new IllegalStateException("entity_config.yaml not found on classpath");
            }
            Map<String, Object> root = yaml.load(is);
            List<Map<String, Object>> entities = (List<Map<String, Object>>) root.get("entities");
            for (Map<String, Object> e : entities) {
                if (entityId.equals(str(e, "entity_id"))) {
                    return new EntityContext(e);
                }
            }
        } catch (java.io.IOException ex) {
            throw new IllegalStateException("Failed to load entity_config.yaml", ex);
        }
        throw new IllegalStateException("No config found for entity: " + entityId);
    }

    public static EntityContext getInstance() { return INSTANCE; }

    // Visible-for-testing factory
    @SuppressWarnings("unchecked")
    public static EntityContext fromMap(Map<String, Object> cfg) {
        return new EntityContext(cfg);
    }

    public String getEntityId() { return entityId; }
    public String getCurrency() { return currency; }
    public String getLocale() { return locale; }
    public String getRegulator() { return regulator; }
    public String getBookingCentre() { return bookingCentre; }
    public BigDecimal getFeeBps() { return feeBps; }
    public BigDecimal getLargePositionThreshold() { return largePositionThreshold; }
    public String getSuitabilityFramework() { return suitabilityFramework; }
    public String getKmsVaultName() { return kmsVaultName; }
    public String getDataRegion() { return dataRegion; }
    public BigDecimal getMinBalance() { return minBalance; }
    public String getDisclosureLocale() { return disclosureLocale; }
    public String getAuditLogPolicy() { return auditLogPolicy; }
    public String getDbConnectionSecret() { return dbConnectionSecret; }
    public String getApiKeySecret() { return apiKeySecret; }

    private static String str(Map<String, Object> m, String k) {
        Object v = m.get(k);
        return v == null ? "" : v.toString();
    }
    private static BigDecimal decimal(Map<String, Object> m, String k) {
        Object v = m.get(k);
        return v == null ? BigDecimal.ZERO : new BigDecimal(v.toString());
    }
}

File: python/entity_context.py
"""Immutable entity context — loaded once per process from YAML."""

import os
from decimal import Decimal
from pathlib import Path

import yaml

_ENTITIES_DIR = Path(__file__).resolve().parent / "entities"


class EntityContext:
    """Immutable configuration holder for a single entity."""

    __slots__ = (
        "entity_id", "currency", "locale", "regulator",
        "booking_centre", "fee_bps", "large_position_threshold",
        "min_balance", "data_region", "kms_vault_name",
        "suitability_framework", "disclosure_locale",
        "audit_log_policy", "db_connection_secret",
    )

    def __init__(self, entity_id: str | None = None) -> None:
        eid = (entity_id or os.environ.get("ACME_ENTITY_ID", "SG")).upper()
        cfg_path = _ENTITIES_DIR / f"{eid}.yaml"
        if not cfg_path.exists():
            raise FileNotFoundError(
                f"No entity config for '{eid}' at {cfg_path}"
            )
        with open(cfg_path, "r") as fh:
            cfg = yaml.safe_load(fh)

        self.entity_id: str = cfg["entity_id"]
        self.currency: str = cfg["currency"]
        self.locale: str = cfg["locale"]
        self.regulator: str = cfg["regulator"]
        self.booking_centre: str = cfg.get("booking_centre", self.entity_id)
        self.fee_bps: Decimal = Decimal(str(cfg["fee_bps"]))
        self.large_position_threshold: Decimal = Decimal(
            str(cfg.get("large_position_threshold", 250000))
        )
        self.min_balance: Decimal = Decimal(
            str(cfg.get("min_balance", 0))
        )
        self.data_region: str = cfg["data_region"]
        self.kms_vault_name: str = cfg["kms_vault_name"]
        self.suitability_framework: str = cfg["suitability_framework"]
        self.disclosure_locale: str = cfg.get(
            "disclosure_locale", self.locale
        )
        self.audit_log_policy: str = cfg.get(
            "audit_log_policy", "standard"
        )
        # Secret reference — NEVER plaintext
        vault = self.kms_vault_name
        self.db_connection_secret: str = (
            f"${{KEYVAULT:{vault}-db-connection-string}}"
        )

    def __setattr__(self, _name: str, _value: object) -> None:
        if hasattr(self, "entity_id"):
            raise AttributeError("EntityContext is immutable")
        super().__setattr__(_name, _value)